src/Service/NoHitService.php

```diff
@@ -8,40 +8,76 @@
 namespace App\Service;
 
 use App\Event\SearchNoHitEvent;
+use App\Utils\Types\NoHitItem;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\HttpKernel\Event\TerminateEvent;
 use Symfony\Component\HttpKernel\KernelEvents;
 
 /**
  * Class NoHitService.
  */
-class NoHitService
+final class NoHitService
 {
-    protected $dispatcher;
+    private $dispatcher;
+    private $metricsService;
+    private $noHitsProcessingEnabled;
 
     /**
      * NoHitService constructor.
      *
      * @param EventDispatcherInterface $dispatcher
+     * @param MetricsService $metricsService
+     * @param bool $noHitsProcessingEnabled
      */
-    public function __construct(EventDispatcherInterface $dispatcher)
+    public function __construct(EventDispatcherInterface $dispatcher, MetricsService $metricsService, bool $noHitsProcessingEnabled)
     {
         $this->dispatcher = $dispatcher;
+        $this->metricsService = $metricsService;
+        $this->noHitsProcessingEnabled = $noHitsProcessingEnabled;
     }
 
     /**
-     * Register no hits.
+     * Send event to register identifiers that gave no search results.
      *
-     * @param array $noHits
+     * @param string $type
+     *   The type ('pid', 'isbn', etc) of identifiers given
+     * @param array $requestIdentifiers
+     *   Array of requested identifiers of {type}
+     * @param array $foundIdentifiers
+     *   Array of identifiers of {type} with covers found
      */
-    public function registerNoHits(array $noHits)
+    public function handleSearchNoHits(string $type, array $requestIdentifiers, array $foundIdentifiers): void
     {
-        $this->dispatcher->addListener(
-            KernelEvents::TERMINATE,
-            function (TerminateEvent $event) use ($noHits) {
-                $noHitEvent = new SearchNoHitEvent($noHits);
-                $this->dispatcher->dispatch($noHitEvent::NAME, $noHitEvent);
+        $notFoundIdentifiers = array_diff($requestIdentifiers, $foundIdentifiers);
+
+        $this->metricsService->counter('no_hit_event_duration_seconds', 'Total number of no-hits', count($notFoundIdentifiers), ['type' => 'rest']);
+        $this->dispatchNoHits($type, $notFoundIdentifiers);
+    }
+
+    /**
+     * Dispatch no hits event.
+     *
+     * @param string $type
+     *   The type ('pid', 'isbn', etc) of identifiers given
+     * @param array $identifiers
+     *   Array of no hit identifiers of {type}
+     */
+    private function dispatchNoHits(string $type, array $identifiers): void
+    {
+        if ($this->noHitsProcessingEnabled && !empty($identifiers)) {
```

Comment:
It might be use full to log that `no hits` was found even they may not be processed. We have missed that data in faktor export and are using a hack to find these in exports currently.